ERCS/erc-7794.md

```diff
@@ -0,0 +1,820 @@
+   */
+  function getDisbursement(
+    bytes32 grantId,
+    uint256 milestoneDate
+  ) external view returns (Disbursements memory);
+
+  /**
+   * @dev Retrieves the list of external links associated with a specific grant
+   * @param grantId The unique identifier of the grant
+   * @return An array of strings representing the external links
+   */
+  function getExternalLinks(
+    bytes32 grantId
+  ) external view returns (string[] memory);
+}
+```
+
+When calling the `registerGrant` function:
+
+- The `grantManager` **MUST** submit a valid grantManager address that is not the zero address.
+- The `community` label **MUST** be a non-empty string.
+- The `grant` ID **MUST** be unique and not already registered in the system.
+
+When editing overall grant details:
+
+- The `grantManager` **MUST** be the current grant manager to make changes to the grant.
+
+When adding a `milestoneDate`:
+
+- The `milestoneDate` **MUST** not exist in the milestonesDates set.
+
+When editing disbursments:
+
+- The `milestoneDate` **MUST** be a valid milestone date associated with the grant.
+
+When adding `externalLinks`:
+
+- The string **MUST** not be empty.
+
+
+## Rationale
+
+The design of this Grant Registry Contract is driven by the need for a flexible and modular system that supports a wide range of grant programs across different chains. The rationale for the key design decisions is outlined below:
+
+1. Separation of Fields: The division of fields into different categories, such as identification, grant data, and disbursement-related information, allows for a more efficient use of on-chain storage. Immutable fields like id, chainid, and community are kept separate from mutable fields, ensuring that core identification elements remain unchanged, while other aspects like milestones and participants can be updated throughout the grant lifecycle.
+
+2. Use of EnumerableSet: By leveraging EnumerableSet for managing participants and milestone dates, the contract allows for dynamic updates, such as changes in team composition or the addition of new milestones. This approach offers flexibility without sacrificing the ability to efficiently track changes.
+
+3. Modular Disbursement Handling: Not all grant programs will choose to perform disbursements on-chain. By allowing disbursements to be managed through external links, the contract remains modular and adaptable to different use cases. Programs that prefer to handle disbursements off-chain can still use the registry for status tracking, ensuring broad applicability across different ecosystems.
+
+4. Gas Optimization: The decision to generate grantId using only the immutable identification fields (with keccak256 and a timestamp) minimizes upfront gas costs. This ensures that essential information is logged efficiently, while the mutable data can be submitted or modified later as the project evolves.
```

Comment:
This looks tied to your reference implementation (unless it's mandatory to use this algorithm to generate grantId?)